### ISSUE TITLE
Hot Reloading is not working

### ISSUE THREAD

**Issue by williamliangwl** [facebook/react-native#18899]
- [ x ] I have reviewed the [documentation](https://facebook.github.io/react-native)
- [ x ] I have searched [existing issues](https://github.com/facebook/react-native/issues)
- [ x ] I am using the [latest React Native version](https://github.com/facebook/react-native/releases)

Hot Reloading is not updating the view in Android device even we have made changes. It happens **after** a Reload triggered from the device.

I have tested this issue in v0.55.2, 0.54.0 and 0.53.3. The only version that has hot reload still functioning well is 0.53.3, this problem **starts** occuring from 0.54.0

## Environment

  OS: Windows 10
  Node: 6.11.5
  Yarn: 1.3.2
  npm: 4.6.1
  Watchman: Not Found
  Xcode: N/A
  Android Studio: Version  3.0.0.0 AI-171.4443003

Packages: (wanted => installed)
  react: 16.3.1 => 16.3.1
  react-native: 0.55.2 => 0.55.2

## Steps to Reproduce

1. Run `react-native init NoHotLoad`
2. Run `react-native run-android`
3. When the project loaded, enable the Hot Reloading
4. Make changes to the App.js, like editing the text. The change **should** appear in the device.
5. Do Reload on the device
6. Repeat step 4, but the change **won't** appear at the device. Just a toast of Hot reloading appears.

**Another scenario**: after installing the app to the device, replace step 2 above with `npm start` should also reproduces this problem.

## Expected Behavior

The Hot Reloading should be able to hot-reload all the changes, regardless number of attempts of reload.

## Actual Behavior

The Hot Reloading only works before a Reload action triggerred

**Comment by gilbertKaradja** [comment]
Same here.

**Comment by gilbertKaradja** [comment]
Im not quite sure what changed but my hot-reload just started working again.

**Comment by DavidNorena** [comment]
Same here man, also had to downgrade to 0.53.3

**Comment by DavidNorena** [comment]
I can confirm neither working for ios

**Comment by williamliangwl** [comment]
@gilbertKaradja do you mean started working even after a Reload happens ? May I know what version are you using?

@DavidNorena thanks for confirming in ios side!

**Comment by cihadturhan** [comment]
I have 3 different projects with versions 0.51, 0.54 and 0.55. The first project works fine.
I think this happened after a they announced new metro (v0.52 I guess)

You can see detailed bug and debug report the link above. I don't know when there will be a fix, meanwhile, you can use the simple dirty hack I did for my project:

https://gist.github.com/cihadturhan/dda54a25eae398d7db0b06292f0cac9c

**Comment by DavidNorena** [comment]
@cihadturhan does it really work, HMR and/or Live Reload, with your hack ?

**Comment by cihadturhan** [comment]
@DavidNorena yeah. Why should I write if it's not working 😄 
Keep in mind that you should work with single emulator/phone

**Comment by DavidNorena** [comment]
Sorry, lol, I haven't test it yet ! lol I write you later and THANKS !

**Comment by amcsi** [comment]
I'm getting this same issue on Windows 10. Although hot loading does work _sometimes_, but even then only for a _short time_ before it just stops actually hot loading (despite the toast "Hot Loading..." appearing).

**Comment by deboyblog** [comment]
Same issue on Windows 10
react-native: 0.55.2
react: 16.3.1

**Comment by DavidNorena** [comment]
@cihadturhan bad luck man, it didn't work for me :(

**Comment by DavidNorena** [comment]
Hi It works with
```
$ create-react-native-app HotLoading
$ cd HotLoading
$ yarn eject
$ yarn android #or yarn ios
```
Note: If you get any RED screen in the emulator run this one:
`$ watchman watch-del-all && react-native start --reset-cache`
and then ctrl -> c to stop the packager.
and again
$ yarn android #or yarn ios

and its working, at least you are not longer stuck at 0.53.3

"react": "16.3.0",
"react-native": "~0.54.1"

**Comment by DavidNorena** [comment]
Because I'm using a lot of native libraries like react-native-navigation, react-native-firebase, and also need typescript support in my project, and the most important (SYMLINK SUPPORT: Metro doesn't have it :( ) and with this and other issues, I ended up using hault, and the [setup](https://callstack.github.io/haul/docs/hmr_setup.html) wasn't that hard, with react-native 0.55.3 in hault HMR its working even after a reload :) I'll give it a chance !!!

**Comment by moshanghan** [comment]
@cihadturhan I can't open your link  https://gist.github.com/cihadturhan/dda54a25eae398d7db0b06292f0cac9c
can you write down here？

**Comment by cihadturhan** [comment]
@moshanghan the link is up. Don't know if it's related to your ISP. Anyway here it is:
Filename: `react-native/local-cli/server/util/attachWebsocketServer.js`
```diff
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @format
 * @flow
 */

'use strict';

import type {Server as HTTPServer} from 'http';
import type {Server as HTTPSServer} from 'https';

type WebsocketServiceInterface<T> = {
  +onClientConnect: (
    url: string,
    sendFn: (data: string) => mixed,
  ) => Promise<T>,
  +onClientDisconnect?: (client: T) => mixed,
  +onClientError?: (client: T, e: Error) => mixed,
  +onClientMessage?: (client: T, message: string) => mixed,
};

type HMROptions<TClient> = {
  httpServer: HTTPServer | HTTPSServer,
  websocketServer: WebsocketServiceInterface<TClient>,
  path: string,
};

/**
 * Attaches a WebSocket based connection to the Packager to expose
 * Hot Module Replacement updates to the simulator.
 */
function attachWebsocketServer<TClient: Object>({
  httpServer,
  websocketServer,
  path,
}: HMROptions<TClient>) {
  const WebSocketServer = require('ws').Server;
  const wss = new WebSocketServer({
    server: httpServer,
    path: path,
  });

+  let oldClient = null;

  wss.on('connection', async ws => {
    let connected = true;
    const url = ws.upgradeReq.url;

    const sendFn = (...args) => {
      if (connected) {
        ws.send(...args);
      }
    };

+    if(oldClient){
+      websocketServer.onClientDisconnect(oldClient);
+    }

    const client = await websocketServer.onClientConnect(url, sendFn);
+   oldClient = client;

    ws.on('error', e => {
      websocketServer.onClientError && websocketServer.onClientError(client, e);
    });

    ws.on('close', () => {
      websocketServer.onClientDisconnect &&
        websocketServer.onClientDisconnect(client);
      connected = false;
    });

    ws.on('message', message => {
      websocketServer.onClientMessage &&
        websocketServer.onClientMessage(client, message);
    });
  });
}

module.exports = attachWebsocketServer;
```

**Comment by moshanghan** [comment]
@cihadturhan it works ，thanks!   In China many links can't be open T_T

**Comment by zeltice** [comment]
This is a show-stopper bug for RN 54+.  Hard to believe there can be people using RN 54 and RN 55 without any hot-reload.

**Comment by FlorianBergmann** [comment]
Is there any word on when this might be fixed? It slows down the development process quite a bit and the issue has existed now for longer than a month. We just updated to the last RN version a few days ago and now we are considering to downgrade again.

**Comment by tribou** [comment]
@cihadturhan's fix works like a charm.

So if you wish to stay on RN 54+, you can perform an unfortunate but helpful hack until this fix drops:

1) Add @cihadturhan's file above to `vendor/attachWebsocketServer.js` (or similar)
1) Add a `package.json` `postinstall` script:

```json
  "scripts": {
    "postinstall": "cp vendor/attachWebsocketServer.js node_modules/react-native/local-cli/server/util/",
```

**Comment by mortezaalizadeh** [comment]
Same problem here on Arch Linux, used to have that before with my Linux Mint and always thought it is an environmental issue. I am running RN 0.55.4

**Comment by mgambati** [comment]
I'm on RN 0.55 and can't make it work.
@cihadturhan fix is not working here.
Any one got another workaround?

**Comment by MWalid** [comment]
Same problem, the fix doesn't work

**Comment by perryiv** [comment]
Using react-native 0.54.4.
The fix worked for me.
Thanks!

**Comment by navata** [comment]
Current, I have used  react-native-0.55.4 and met same this issue.
When I saved my code, I must re-open screen or reload application for updating UI.

**Comment by skptricks** [comment]
Refer to this article  : https://www.skptricks.com/2018/07/enable-hot-reloading-in-react-native.html

**Comment by lmichailian** [comment]
Same here using  react-native-0.55.4

**Comment by skptricks** [comment]
It is very simple. Just sharing link, it will help you to enable hot
reloading in react natve.

Link :
https://www.skptricks.com/2018/07/enable-hot-reloading-in-react-native.html

**Comment by williamliangwl** [comment]
@navata @lmichailian have you guys tried cihadturan's solution [above](https://github.com/facebook/react-native/issues/18899#issuecomment-387409033)?

@skptricks have you tried to reproduce this issue by following the steps specified? if you manage to not having any problem on Hot Reloading after following the steps, we are happy to know your React Native version and steps you have done to solve the issue

**Comment by skptricks** [comment]
Currently using react-native@0.55.4 version.

**Comment by skptricks** [comment]
1. First uninstall the application from emulator.
2. Then run the react native application in emulator and enable the hot reloading.

**Comment by lustigdev** [comment]
Found a fix for me.

[Video with all details for those interested](https://www.youtube.com/watch?v=sdQCGab4MjY)

### Fix
1) Update to 0.55.4
To do that, navigate to the root directory of your project and enter the line
    npm install react-native@0.55.4
    rm -rf node_modules
    npm i

2) Uninstall the application and kill the package manager

3) Run your project again to make sure that upgrading the react-native version worked properly

4) If you render any code with JS6 style functions, change them to traditional style. I cover this in the video if you don't know what I mean.

    foo = () => {
        // This doesn't work with hot reload on v.0.55.4
    }

    foo() {
        // This does work with hot reload on v.0.55.4
    }

Hopefully that helps. No other changes required!

**Comment by skptricks** [comment]
Check out below link to enable hot reloading in react native application :
Enable Hot Reloading In React Native
<https://www.skptricks.com/2018/07/enable-hot-reloading-in-react-native.html>

**Comment by QShengW** [comment]
0.54.4 Perfect solution by using the method of cihadturhan Thanks

**Comment by osdiab** [comment]
I'm running into this on `0.56`, though the file that @cihadturhan refers to seems to have been moved to the `metro` package; not sure how this patch ought to be used there.

Is this the issue tracking the underlying Metro problem? https://github.com/facebook/metro/issues/165

Hope this gets resolved legitimately sometime rather than requiring patching it like this.

**Comment by BenBtg** [comment]
Same issue here. HotReload only working till a redbox error hit. Then reload forced when redbox issue fixed and HotReload stops working till a forced reload. Can repro with a clean react-native init app.

**Comment by Phecda** [comment]
Hi, for anyone using Typescript, try changing your `tsconfig.json` like this:
```javascript
"compilerOptions": {
    "target": "es2015",
    "module": "es2015",
    "moduleResolution": "node",

    //  comment next line
    // "resolveJsonModule": true
}
```
See [this article](https://medium.com/@pavel_55299/typescript-react-native-hot-reload-not-working-5175c6506eb2)

In this case I have to abandon importing from a `json` file. 

I'm using RN 0.55.3.

**Comment by Batuu13** [comment]
Same issue here. Fresh start works hit hot-reload until I have a redbox error. Then you need to either redeploy or kill the app and manual restart.

**Comment by P48L0** [comment]
> I have 3 different projects with versions 0.51, 0.54 and 0.55. The first project works fine.
> I think this happened after a they announced new metro (v0.52 I guess)
> 
> You can see detailed bug and debug report the link above. I don't know when there will be a fix, meanwhile, you can use the simple dirty hack I did for my project:
> 
> https://gist.github.com/cihadturhan/dda54a25eae398d7db0b06292f0cac9c

Thanks dude!! your solution fixed my issue. I was experiencing a 20+ seconds delay in seeing the HotLoading on my phone, now i see it almost instantly as soon as i save.

**Comment by khorark** [comment]
>  I have 3 different projects with versions 0.51, 0.54 and 0.55. The first project works fine.
>  I think this happened after a they announced new metro (v0.52 I guess)
>  You can see detailed bug and debug report the link above. I don't know when there will be a fix, meanwhile, you can use the simple dirty hack I did for my project:
>  https://gist.github.com/cihadturhan/dda54a25eae398d7db0b06292f0cac9c

It's not work for RN 0.57, because file attachWebsocketServer.js not exist. 

Maybe, some have solution for this problem?

**Comment by jan-happy** [comment]
> It's not work for RN 0.57, because file attachWebsocketServer.js is not exist.
> 
> Maybe, some have solution for this problem?
For me patching it in node_modules/metro/src/lib/attachWebsocketServer.js.flow worked. 👍

**Comment by kida7** [comment]
> > It's not work for RN 0.57, because file attachWebsocketServer.js is not exist.
> > Maybe, some have solution for this problem?
> 
> For me patching it in node_modules/metro/src/lib/attachWebsocketServer.js.flow worked. 👍

I patched following @cihadturhan 's instruction and yours but  there is a error threw:
```
import type { Server as HTTPServer } from 'http';
^^^^^^

SyntaxError: Unexpected token import
    at createScript (vm.js:80:10)
    at Object.runInThisContext (vm.js:139:10)
    at Module._compile (module.js:616:28)
    at Module._compile (/Users/vinhle/Desktop/Projects/vnMov/node_modules/pirates/lib/index.js:83:24)
    at Module._extensions..js (module.js:663:10)
    at Object.newLoader [as .js] (/Users/vinhle/Desktop/Projects/vnMov/node_modules/pirates/lib/index.js:88:7)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)
    at Module.require (module.js:596:17)
```

**Comment by jamessawyer** [comment]
same as React native 0.57.2 I have to use live reload to reload entire project, it's annoying

**Comment by skptricks** [comment]
Switch back to old one : react-native init <projectname> --version
react-native@0.55.4 .

**Comment by BenBtg** [comment]
How is this still an issue? Am I missing something or is hot reload not important feature of RN?

**Comment by kipertech** [comment]
> How is this still an issue? Am I missing something or is hot reload not important feature of RN?

It is super crucial and to me it's a killing feature compared to other frameworks/libraries. It speeds up development a lot.

**Comment by fogil** [comment]
How to fix hot reloading:
1.) DO NOT attempt to upgrade to 0.57.04 -- It will fail!
2.) Start a brand new react-native 0.57.04 project.

**Comment by bntzio** [comment]
@fogil I created a brand new react-native 0.57.4 project and it still doesn't work

**Comment by underbluewaters** [comment]
Hot reloading only seems to work for me with remote debugging on.

**Comment by abhijit-padhy** [comment]
Try adding -- --reset-cache to your run command. 
*for Linux OS

Ref: 
[](https://www.reddit.com/r/reactnative/comments/8mlelt/hot_reloadinglive_reload_stopped_working_how_do_i/)

**Comment by Frexuz** [comment]
How can this get screwed up? It's promised out-of-the-box, as we all expect it to be, and work. 🤦‍♂️ 
0.57.5

**Comment by adilhaddaoui** [comment]
On react-native: 0.57.4, and the same error still occurs, it's really a headache, Any Fix out there ?
whenever I change something in code I need to kill the app and re-run for the code to take place. HRM isn't working at all.

**Comment by fanyer** [comment]
same here on react-native@0.57.4, the experience really sucks.

**Comment by SRandazzo** [comment]
> Hot reloading only seems to work for me with remote debugging on.

Turning on remote debugging also resolves the issue for me, but just as a workaround

**Comment by hackerhgl** [comment]
There isn't "attachWebsocketServer.js" in react-native@0.57.5 so good bye to the patch by @cihadturhan.
Why does the team even bother releasing the new versions If it breaks the existing core features.

**Comment by SRandazzo** [comment]
Just spotted this fix here merged to master: https://github.com/facebook/react-native/pull/22412

:-D

**Comment by s123121** [comment]
confirm working after upgrading to RN 0.57.6 (however, I have to configure the project against for it to work with xcode 10)

https://github.com/facebook/react-native/issues/20774

**Comment by sryze** [comment]
I noticed the same problem, I'm using 0.57.6 on iOS simulator. I think it used to work in 0.57.2.

**Comment by vicke4** [comment]
@cihadturhan's patch is not wokring. Even, @jan-happy's fix didn't help. Using version 0.57.6

**Comment by yikkok-yong** [comment]
Upgrading from 0.48 to 0.57. Hot reloading not working. Tried above steps, nothing is work.
Deleted package-lock.json. `npm install`, regenerate package-lock.json. Hot reloading working.

**Comment by vjtc0n** [comment]
0.57.7 still faces this issue...
Any workaround for this patch?
Thanks.

**Comment by eduplus** [comment]
For me using XCode 10 and 0.57.7: Hot reload enabled if I make change - I can see Hot reloading bar on top and after a while app is just closed (probably crashes). For me hot reload works if debug is running at the same time.

**Comment by timhilgenstoehler** [comment]
> For me using XCode 10 and 0.57.7: Hot reload enabled if I make change - I can see Hot reloading bar on top and after a while app is just closed (probably crashes). For me hot reload works if debug is running at the same time.

Jeah thanks for that! Its working when the Debugger is running!

**Comment by msiripan** [comment]
Recently my project got performance improvement by changing from React.Component to React.PureComponent so the i got an HMR toast in android but the change didn't reflect in the UI.
Solved by following this in SO: https://stackoverflow.com/questions/46013170/react-native-0-47-1-hot-reload-is-not-reflecting-code-changes-on-macos

**Comment by mpraki** [comment]
In my case, the debug server(from Dev Settings) which I had set in the phone got reset when I turned off the Wi-Fi. So, in the debugger menu, I had to choose Dev Settings -> Debug Server and I added back the PC IP and port as 192.168.1.3:8081. It worked for me.

So, if it's not working, one option we can look for is the Debug Server setting.

**Comment by kristerkari** [comment]
> We managed to get it working on react-native@0.57.8 and react@16.6.3 by pinning react-deep-force-update to 1.1.2

I can confirm that this fixed the problem for me when upgrading my project from React Native 0.56.0 to 0.57.8. 

My `yarn.lock` had `react-deep-force-update` locked to version `1.1.1`. 

I manually removed the `react-deep-force-update` from my `yarn.lock` and ran `yarn install` after that. Then when I restarted the packager using `yarn start --reset-cache` hot loading started working again.

**Comment by nicutimofte** [comment]
After updating to `0.58.3` I'm still facing this issue on iOS, but only when debugger is enabled.

**Comment by R4meau** [comment]
I can confirm, still happening in `0.58.4` for Android. Even if I Stop Remote JS Debugging from the dev menu.

**Comment by Joey33** [comment]
So far, I've had to use a workaround - enable 'debug JS remotely'
But adding resolutions like @sbycrosz suggested works like a charm - thanks! :)

**Comment by jedt** [comment]
Hot reloading might NOT work If you don’t use ES6 and still using create-react-class. I tested this by creating a new project with git init and hot-reloading worked ok. Then I modified the example code to something like this

```
var createReactClass = require('create-react-class');
var Greeting = createReactClass({
  render: function() {
    return <h1>Hello, {this.props.name}</h1>;
  }
});
 ```

And hot reloading broke. You might need to downgrade to 0.51 or refactor your codebase. Hope this helps

**Comment by ParhamZare** [comment]
I runned below commands and then its worked
```
watchman watch-del-all
rm -rf $TMPDIR/react-*
rm -rf $TMPDIR/metro-*
rm -rf $TMPDIR/haste-*
npm start -- --reset-cache
``` 
but i think another issue is Use Stateless component with HOC Component 
I wrote below code 
```
const wrapper = () => {
    return (
            <Provider store={store}>
                <App/>
            </Provider>
    );
}

const ReachApp = codePush({
    checkFrequency: codePush.CheckFrequency.ON_APP_RESUME,
    installMode: codePush.InstallMode.IMMEDIATE
})(Wrapper);

AppRegistry.registerComponent('firstApp', () => ReachApp)
```
and then changed code to 

```
class Wrapper extends React.Component {
    constructor(props) {
        super(props)
    }

    render() {
        return (
            <NetworkProvider pingTimeout={2000}>
                <Provider store={store}>
                    <App/>
                </Provider>
            </NetworkProvider>
        )
    }
}

const ReachApp = codePush({
    checkFrequency: codePush.CheckFrequency.ON_APP_RESUME,
    installMode: codePush.InstallMode.IMMEDIATE
})(Wrapper);

AppRegistry.registerComponent('firstApp', () => ReachApp)

```
and its worked in react-native 0.57.0 and react 16.6.3

**Comment by cihati** [comment]
Confirming that @sbycrosz 's fix is working on 0.58.5 with iOS simulator!

**Comment by gideaoms** [comment]
react-native start --reset-cache --max-workers=1

**Comment by raosan** [comment]
if non of above solution works, check your component life cycle method.

I got the problem because i use async on it

`componentDidMount = async () => {...}`

when i change back without async, hot reloading works fine

**Comment by a-eid** [comment]
deleting `.git/index.lock` might help

**Comment by JohannesKlauss** [comment]
A combination of @gideaoms and @sbycrosz  answers did it for me (`react-native@0.58.6`)

**Comment by jan-happy** [comment]
Neither the single solutions or a combination of @gideaoms and @sbycrosz answers work for me (react-native@0.58.6).
I also tried all the other suggestions from above. (rm node_modules, remote debugging, etc)
Now the screen stays black after I fixed an error that caused a red screen.
Only reload helps.

**Comment by sultanularefin** [comment]
None of the above worked for my project, I was using react-native: "0.58.4" then upgraded it to ""0.58.6". But the problem persists.

**Comment by sultanularefin** [comment]
I was facing this problem in my office **windows 10** PC. My code changes don’t reflect in my physical device and in emulator after “Reload” from developer menu (Ctrl +M or with the "Shake Gesture" option). Then I reached this page to find a solution. I have tried many of them but it didn’t work for me.

Later I went my home and pull the remote in my home’s **Ubuntu PC**. Then “Reload” option reflects to my code change. 

Today I came to office and create a new folder, where I pulled my branch from remote and now “Reload option from developer menu” reflects to my code changes.

**I think this issue has something to do with Windows and git.**

Before running this successful build, which reflects to code changes, I did these things as listed below: 

1. Create a new folder.
2. Git clone “my repository”.
3. Git checkout “my branch”
4. From **Folder properties** Uncheck the attribute “Read only (only applies to files in folder)” in **General tab.**
5. In **Security Tab** give every user names and group full control by checking the Checkbox fields.

**Comment by jan-happy** [comment]
React Native 0.59 still has the same issue for me.

**Comment by Inveth** [comment]
Same on newest one for me too - looks like it hangs ONLY after you type in invalid code - anything valid will be reloaded but once node pickup the invalid code it will not hot-reload it anymore.

**Comment by suzilxptr** [comment]
@Inveth agree with you, having the same scenario

**Comment by lettucemode** [comment]
I followed the steps above when running in the emulator, but it didn't seem to work for me. 

Then I tried it on my real android phone. In the react native app, I opened the developer menu and tapped "Enable Hot Loading". It gave me an error message about how the server couldn't be found or used or something, and said that if the device and the dev computer were on the same wireless network, to enter the IP and port in the dev menu's Dev Settings -> Debug Server Host & Port For Device. So I did that (192.168.1.6:8081) and hot reloading is now working on my real phone!

Probably need to do something similar when running through the emulator, but haven't investigated yet.

**Comment by nahidmbstu** [comment]
In my case , Hot reload stops working after 3-4 time i update my code . then i need to reload it manually to see my code update.

**Comment by jsslai** [comment]
I'm not sure if this is a know issue or if this is related but I had to change root component to be class component instead of function component.

**Comment by cdreier** [comment]
i can confirm, if the root component is a class, its working fine for me

```jsx
import React from 'react'
import { createStackNavigator, createAppContainer } from 'react-navigation'
import HomeScreen from './src/Home'

const AppNavigator = createStackNavigator({
  Home: {
    screen: HomeScreen,
  },
  initialRouteName: 'Home',
})

const AppContainer = createAppContainer(AppNavigator)
// no hot loading
// export default AppContainer

class AppNavigatorWrapper extends React.Component {
  render() {
    return <AppContainer />
  }
}
// hot loading is working!
export default AppNavigatorWrapper
```

**Comment by ctrljs** [comment]
@nahidmbstu , I agree , i quote you "In my case , Hot reload stops working after 3-4 time i update my code . then i need to reload it manually to see my code update." ... 

Please can someone give a proper solution to this , it is so time wasting.

**Comment by suzilxptr** [comment]
For me, the problem was expo was caching on the phone. Once I cleared it, it started working. Also, check your network configuration for your mobile and machine.

**Comment by cpojer** [comment]
There are many known problems with Hot Reloading which are existing issues with the hot reloading system separate from React Native. @gaearon was interested in fixing this but I'm going to close this PR here as we can't really do anything about it from within React Native.

**Comment by jan-happy** [comment]
> I'm not sure if this is a know issue or if this is related but I had to change root component to be class component instead of function component.

Does not help here :(

**Comment by gideaoms** [comment]
one of the hot reload issues is related to visual studio code, if you are working on a project and for some reason, you click reload visual studio code, then hot reload will stop working. My tip is: reopen visual studio code every time you install a new extension or when you want to open another project.

**Comment by jan-happy** [comment]
@gideaoms That's a great hint, thank you! 
It indeed works for me! I just wonder: Why?
I don't use Visual Studio for debugging or running the app, I just use it as an editor. How does this affect HMR?

**Comment by mlg87** [comment]
@gideaoms suggestion along with killing metro and starting again (with `--reset-cache`) got hot reloading working for me again

**Comment by gideaoms** [comment]
@jan-happy I'm sorry, but I did not investigate the reason in depth. In fact, I found out after researching a lot. I realized that reload was working with other editors, but it did not work with VSCode. After some tests, I discovered that opening another project without closing the VSCode or reloading the VSCode after installing an extension was breaking the Hot Reloading.

**Comment by pugson** [comment]
> For me, the problem was expo was caching on the phone. Once I cleared it, it started working. Also, check your network configuration for your mobile and machine.

@suzilxptr Can you expand on this?

**Comment by anshgupta1234** [comment]
I've just started using React Native, and hot reloading worked like a charm for 2 minutes and went back to being annoying again. Does anybody know the Hot Reload author(s) so they can help us out?

**Comment by appa3151** [comment]
> if non of above solution works, check your component life cycle method.
> 
> I got the problem because i use async on it
> 
> `componentDidMount = async () => {...}`
> 
> when i change back without async, hot reloading works fine

thanks bor..

**Comment by joewired** [comment]
watchman watch-del <path-to-project-directory>
watchman watch-project <path-to-project-directory>

And rebuilding the project fixed for me.

**Comment by rewieer** [comment]
Got it fixed using 
```
"resolutions": {
  "react-deep-force-update": "1.1.2"
}
```
Into `package.json` then running `yarn`. You can remove it afterwards. Thanks @sbycrosz !

**Comment by TaylorDale** [comment]
@cdreier thanks for that, mine works after wrapping the root component

**Comment by Nsengiyunva** [comment]
I have tried almost all suggested answers. `yarn start --reset-cache` to run packager. I removed the `node modules` folder and did a `yarn install` again. I tried @rewieer answer. None of these seem to work for me. I am using `react native 0.59.8`.
I closed all the other `vscode`  windows, then run the packager from `git bash` terminal and installing too not the `vs code` terminal but hot reloading just wont work.
This bug is still not solved.

**Comment by gaearon** [comment]
I'll reopen since I'm working on a new implementation. However I don't know if the problems I'm fixing are related to the problems in this thread. Some of them might be, some of them probably aren't.

I'll close this after I land the new implementation. If you have a problem after the release that contains it, it would be best to create a new issue with a concrete reproducible example and/or video.

For now let's keep it open.

**Comment by tpisto** [comment]
@gaearon Would the new HMR work also with the good old create-react-class?

**Comment by gaearon** [comment]
Yes (but it would remount the edited component). In the new implementation, we only support preserving state for the edited components for functions with Hooks. Supporting it for classes without subtly altering semantics and making the feature fragile is too difficult.

Btw I'd appreciate if people could start making small isolated example projects where hot reloading doesn't work. Like maybe you use some navigation library where it breaks. They would be helpful so I can debug them.

**Comment by EduVencovsky** [comment]
In android, when I press `Ctrl + S` it shows the android toast with `Hot Loading...` But it doesn't change anything. Only when I manually reload it (and the green bar on the top appears), it changes.
 
When I enable live reloading, one weird thing happens. In the metro bundler, after the first loading, it always shows `100.0% (1/1), done.` like if there are no changes.

I'm using react-native 0.59.3

**Comment by chuttam** [comment]
@gaearon - Here is a tiny project (essentially the result of `react-native init`) that replicates the issue.

https://github.com/chuttam/react-native-hotreload-issue18899

The project's README has specific instructions for replicating the Hot Reloading malfunction. As @jsslai  and @cdreier confirmed, changing the root component from a class to a function component triggers the issue.

**Comment by gaearon** [comment]
The new implementation has landed! We're calling it Fast Refresh.
You can see a demo here: https://mobile.twitter.com/reactnative/status/1144629612921720833

I'll leave this issue open until I verify that it indeed solves the reproduction cases posted in this thread.

---

@EduVencovsky Sorry, this isn't helpful without a reproducing project.

@chuttam Thanks for a repro case. I will check whether my changes fix it. (I expect them to).

If anyone has more reproducing cases, now is the time to post them.

**Comment by gaearon** [comment]
**If you have any concrete cases that are still broken please post reproducing examples with instructions in this thread. This is your chance to see them fixed.**

**Comment by RobertPaul01** [comment]
I'm not sure if this case has been addressed, but I have encountered problems with using base component classes. This happens if you create a `BaseComponent` that extends `Component` or `PureComponent` and then extend that `BaseComponent` to create your screens. With hot reloading enabled, making a changing and then saving will pop the `Hot Loading...` message, but the content will not update.

Here is an example: https://github.com/RobertPaul01/HotReloadTest/blob/master/App.js

**Comment by gaearon** [comment]
Just checked that this is fixed.

(But you probably should avoid creating component inheritance hierarchies.)

**Comment by cdreier** [comment]
Hi, i just tested with 0.60.0-rc.3 (i think this is the correct version?), and still having the same issue as described in my comment above

https://github.com/facebook/react-native/issues/18899#issuecomment-475414525

Repo: https://github.com/cdreier/rn-fast-refresh/blob/master/App.tsx

This is not a big issue, with a class component as root everything is working very well. But if this intended, it should appear somewhere in the docs?

**Comment by gaearon** [comment]
@cdreier 0.60 is *not* going to contain any of these fixes, sorry. They are slated for 0.61.

**Comment by nathsimpson** [comment]
Also, make sure you don't have an `index.lock` file in your .git directory :) Deleting it could solve your issue.

**Comment by AMSTKO** [comment]
I'm new to react native and I have this Hot Realoding issue with a "fresh" stack and the example project
You can see a little gif of the behavor bellow

```
➜  AwesomeProject react-native info
info Fetching system and libraries information...
System:
    OS: macOS 10.14.5
    CPU: (4) x64 Intel(R) Core(TM) i7-7660U CPU @ 2.50GHz
    Memory: 233.45 MB / 16.00 GB
    Shell: 5.3 - /bin/zsh
  Binaries:
    Node: 10.16.0 - /usr/local/bin/node
    Yarn: 1.17.3 - /usr/local/bin/yarn
    npm: 6.9.0 - /usr/local/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  SDKs:
    iOS SDK:
      Platforms: iOS 12.2, macOS 10.14, tvOS 12.2, watchOS 5.2
    Android SDK:
      API Levels: 28, 29
      Build Tools: 28.0.3, 29.0.1
      System Images: android-28 | Intel x86 Atom_64, android-28 | Google APIs Intel x86 Atom, android-29 | Google APIs Intel x86 Atom
  IDEs:
    Android Studio: 3.4 AI-183.6156.11.34.5692245
    Xcode: 10.2.1/10E1001 - /usr/bin/xcodebuild
  npmPackages:
    react: 16.8.6 => 16.8.6
    react-native: 0.60.3 => 0.60.3
  npmGlobalPackages:
    react-native-cli: 2.0.1
```

![rn_off](https://user-images.githubusercontent.com/408968/61368429-1d2fb900-a88e-11e9-99a9-9ef4aa352e25.gif)

**Comment by aminy** [comment]
> In android, when I press `Ctrl + S` it shows the android toast with `Hot Loading...` But it doesn't change anything. Only when I manually reload it (and the green bar on the top appears), it changes.
> 
> When I enable live reloading, one weird thing happens. In the metro bundler, after the first loading, it always shows `100.0% (1/1), done.` like if there are no changes.
> 
> I'm using react-native 0.59.3

I have the exact same issue on 0.59.10 
any solutions or workarounds?

**Comment by femartin** [comment]
I just uninstalled EXPO from my Android device and reinstall it and now is working again ok.

**Comment by DaniAkash** [comment]
Is there any test versions available for me to check if fast refresh can solve the issue in my project? @gaearon

**Comment by rewieer** [comment]
Just installed RC 0.61.0 with FastRefresh.
Before, functional component would reset the entire navigator. Now it just refreshes the component.
A huge win for me, thanks @gaearon !

**Comment by GreatAuk** [comment]
> Just installed RC 0.61.0 with FastRefresh.
> Before, functional component would reset the entire navigator. Now it just refreshes the component.
> A huge win for me, thanks @gaearon !

seems 0.62-rc.0 not working。

**Comment by gaearon** [comment]
Known issues with hot reloading have ben fixed in 0.61.
The feature is a complete rewrite and is now called "Fast Refresh".

The 0.61 RC0 still has some issues (redbox not appearing) that will be fixed in the final 0.61 release.

I'm going to close and lock this issue to prevent further confusion.
**If you have issues in React Native 0.61 with Fast Refresh, please file a new issue and ping me.**

**Comment by gaearon** [comment]
You can try [React Native 0.61 RC 3](https://twitter.com/grabbou/status/1171472333313175552) which should be representative of what will go out as stable.
